File: android/app/src/main/java/technology/tiny/app/fleet/Media.kt
package technology.tiny.app.fleet

import android.app.ActivityManager
import android.content.Context
import android.content.Intent
import android.net.Uri

/**
 * Spotify / media deep-links (iOS Media.swift parity). A "play … on spotify"
 * relay prompt is a REAL device action, not a chat proxy: the phone opens the
 * Spotify search deep-link so the track/playlist is one tap from playing.
 *
 * [spotifyLinks] extracts open.spotify.com URLs the agent puts in a reply so the
 * chat can render them as tappable chips (mirrors iOS Media.spotifyLinks(in:)).
 * [musicQuery] strips the "play"/"on spotify" scaffolding to the bare query.
 * [searchUrl] builds the search deep-link. [open] launches it, but only when
 * the app is foreground — Android forbids background activity launches (10+),
 * the same constraint iOS enforces via applicationState == .active.
 */
object Media {
    private val LINK = Regex("""https://open\.spotify\.com/[A-Za-z0-9/_?=&.-]+""")

    /** All open.spotify.com URLs in a block of text, in order, de-duplicated. */
    fun spotifyLinks(text: String): List<String> =
        LINK.findAll(text).map { it.value }.distinct().toList()

    // Trailing/leading noise trimmed off a resolved query — whitespace plus the
    // punctuation iOS strips via .punctuationCharacters (so "play X!" → "X"). Only
    // the ENDS are trimmed, so an interior "AC/DC" or "Sgt. Pepper" is untouched.
    private const val EDGE_NOISE = "!?.,;:…\"'`()[]{}-–—"

    /**
     * Strip the play-verb and spotify/phone suffixes so what's left is the bare
     * thing to search for. "play daft punk on spotify" → "daft punk".
     *
     * iOS Media.musicQuery matches `\bplay\s` — the verb WHEREVER it leads, not just
     * at the string start — so a conversational relay prompt ("can you play daft punk
     * on spotify") resolves to "daft punk", not "can you play daft punk". The word
     * boundary keeps "display the charts" from being mistaken for the command.
     */
    fun musicQuery(prompt: String): String {
        var q = prompt
        // Take everything after the first real "play " verb (word-boundary anchored).
        Regex("""\bplay\s""", RegexOption.IGNORE_CASE).find(q)?.let { q = q.substring(it.range.last + 1) }
        // Repeatedly drop trailing scaffolding (`while`, iOS parity — "… on spotify spotify").
        for (suffix in listOf(" on spotify", " in spotify", " on the phone", " spotify")) {
            while (q.endsWith(suffix, ignoreCase = true)) q = q.dropLast(suffix.length)
        }
        return q.trim { it.isWhitespace() || it in EDGE_NOISE }
    }

    /** Spotify search deep-link for a query (null/blank → the plain search page). */
    fun searchUrl(query: String): String {
        val q = query.trim()
        if (q.isEmpty()) return "https://open.spotify.com/search"
        return "https://open.spotify.com/search/" + Uri.encode(q)
    }

    /**
     * Open a Spotify URL on the phone — ONLY when the app is foreground. Returns
     * true if launched, false if backgrounded (caller hands back the link instead).
     * Android blocks background activity launches; iOS blocks the same via
     * applicationState. FLAG_ACTIVITY_NEW_TASK because the launch originates from
     * the relay service context, not an Activity.
     */
    fun open(context: Context, url: String): Boolean {
        if (!isForeground(context)) return false
        val uri = runCatching { Uri.parse(url) }.getOrNull() ?: return false
        return runCatching {
            context.startActivity(Intent(Intent.ACTION_VIEW, uri).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK))
            true
        }.getOrDefault(false)
    }

    private fun isForeground(context: Context): Boolean {
        val am = context.getSystemService(Context.ACTIVITY_SERVICE) as? ActivityManager ?: return false
        return am.runningAppProcesses?.any {
            it.processName == context.packageName &&
                it.importance == ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND
        } ?: false
    }
}

File: ios/Tiny/Sources/Media.swift
/**
 * Media — Spotify hooks, URL-scheme-first (MusicKit later, per backlog).
 *
 *   - spotifyLinks(in:): open.spotify.com URLs in assistant replies become
 *     tappable "Open in Spotify" chips under the bubble
 *   - searchURL(_:): universal link that deep-links into the Spotify app
 *     when installed, web player otherwise — used by the relay so the web
 *     agent can say "play X on my phone" (foreground only; iOS forbids
 *     opening apps from a backgrounded process)
 */
import Foundation

enum Media {
    static func spotifyLinks(in text: String) -> [URL] {
        let pattern = "https://open\\.spotify\\.com/[A-Za-z0-9/_?=&.-]+"
        guard let re = try? NSRegularExpression(pattern: pattern) else { return [] }
        let range = NSRange(text.startIndex..., in: text)
        var seen = Set<String>()
        return re.matches(in: text, range: range).compactMap {
            guard let r = Range($0.range, in: text) else { return nil }
            let s = String(text[r])
            guard seen.insert(s).inserted else { return nil }
            return URL(string: s)
        }
    }

    static func searchURL(_ query: String) -> URL? {
        let q = query.addingPercentEncoding(withAllowedCharacters: .urlPathAllowedCharacterSetSafe) ?? query
        return URL(string: "https://open.spotify.com/search/\(q)")
    }

    /// "play daft punk on spotify" → "daft punk"
    static func musicQuery(from prompt: String) -> String {
        var q = prompt
        // Word-boundary match so the "play" verb is stripped wherever it leads
        // ("play …", "can you play …") WITHOUT matching it as a substring —
        // \b means "dis[play] the charts" isn't mistaken for the command.
        if let r = q.range(of: "\\bplay\\s", options: [.regularExpression, .caseInsensitive]) {
            q = String(q[r.upperBound...])
        }
        for suffix in [" on spotify", " in spotify", " on the phone", " spotify"] {
            while q.lowercased().hasSuffix(suffix) { q = String(q.dropLast(suffix.count)) }
        }
        return q.trimmingCharacters(in: .whitespacesAndNewlines.union(.punctuationCharacters))
    }
}

private extension CharacterSet {
    /// urlPathAllowed still permits "/" — percent-encode it inside a search term
    static var urlPathAllowedCharacterSetSafe: CharacterSet {
        var set = CharacterSet.urlPathAllowed
        set.remove(charactersIn: "/?")
        return set
    }
}
